src/tests/ftest/launch.py

```diff
@@ -1459,886 +1108,1972 @@ def ignore_unknown(self, node):  # pylint: disable=no-self-use,unused-argument
-    local_host = NodeSet(socket.gethostname().split(".")[0])
-    yaml_hosts = NodeSet()
-    if args.include_localhost and key_match != YAML_KEYS["test_servers"]:
-        yaml_hosts.add(local_host)
-    found_client_key = False
-    for key, value in list(find_yaml_hosts(test_yaml).items()):
-        display(args, "  Found {}: {}".format(key, value))
-        if key_match is None or key == key_match:
-            display(args, "    Adding {}".format(value))
-            if isinstance(value, list):
-                yaml_hosts.add(NodeSet.fromlist(value))
+    status = {
+        "status": 0,
+        "stop": NodeSet(),
+        "disable": NodeSet(),
+        "reset-failed": NodeSet()}
+    status_states = {
+        "stop": ["active", "activating", "deactivating"],
+        "disable": ["active", "activating", "deactivating"],
+        "reset-failed": ["failed"]}
+    command = ["systemctl", "is-active", service]
+    result = run_remote(log, hosts, " ".join(command))
+    for data in result.output:
+        if data.timeout:
+            status["status"] = 512
+            status["stop"].add(data.hosts)
+            status["disable"].add(data.hosts)
+            status["reset-failed"].add(data.hosts)
+            log.debug("  %s: TIMEOUT", data.hosts)
+            break
+        # log.debug(" %s: %s", data.hosts, "\n".join(data.stdout))
+        for key, state_list in status_states.items():
+            for line in data.stdout:
+                if line in state_list:
+                    status[key].add(data.hosts)
+                    break
+    return status
+
+
+def reset_server_storage(log, test):
+    """Reset the server storage for the hosts that ran servers in the test.
+
+    This is a workaround to enable binding devices back to nvme or vfio-pci after they are unbound
+    from vfio-pci to nvme.  This should resolve the "NVMe not found" error seen when attempting to
+    start daos engines in the test.
+
+    Args:
+        log (logger): logger for the messages produced by this method
+        test (TestInfo): the test information
+
+    Returns:
+        int: status code: 0 = success, 512 = failure
+
+    """
+    hosts = test.hosts.servers
+    log.debug("-" * 80)
+    log.debug("Resetting server storage after running %s", test)
+    if hosts:
+        commands = [
+            "if lspci | grep -i nvme",
+            "then daos_server storage prepare -n --reset && "
+            "sudo rmmod vfio_pci && sudo modprobe vfio_pci",
+            "fi"]
+        log.info("Resetting server storage on %s after running '%s'", hosts, test)
+        result = run_remote(log, hosts, f"bash -c '{';'.join(commands)}'", timeout=600)
+        if not result.passed:
+            log.debug("Ignoring any errors from these workaround commands")
+    else:
+        log.debug("  Skipping resetting server storage - no server hosts")
+    return 0
+
+
+class LaunchException(Exception):
+    """Exception for launch.py execution."""
```

Comment:
Should this be defined higher up in the file?

Reply:
Done